File: solution/leetcode/2.cpp
// https://leetcode.com/problems/add-two-numbers/
// Standard `Linked List` based solution

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        if (l1 == nullptr && l2 == nullptr) return nullptr;
        
        ListNode* head = new ListNode(0);
        ListNode* node = head;
        
        ListNode* first = l1;
        ListNode* second = l2;
        int carry = 0;
        
        while (first != nullptr && second != nullptr) {
            int sum = first->val + second->val + carry;
            node->next = new ListNode(sum%10);
            node = node->next;
            carry = sum/10;
            first = first->next; second = second->next;
        }
        
        if (second != nullptr) swap(first, second);
        
        while (first != nullptr) {
            int sum = first->val + carry;
            node->next = new ListNode(sum%10);
            node = node->next;
            carry = sum/10;
            first = first->next;
        }
        
        if (carry) {
            node->next = new ListNode(carry);
            carry = 0;
            node = node->next;
        }
        
        head = head->next;
        return head;
    }
};

File: solution/leetcode/2.py
# https://leetcode.com/problems/add-two-numbers/description/

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        carry = 0
        head = ListNode()
        cur = head

        while l1 and l2:
            sum = l1.val + l2.val + carry
            carry = sum // 10
            digit = sum % 10
            cur.next = ListNode(digit)
            l1, l2 = l1.next, l2.next
            cur = cur.next
        
        if not l1:
            l1 = l2
        while l1:
            sum = l1.val + carry
            carry = sum // 10
            digit = sum % 10
            cur.next = ListNode(digit)
            l1 = l1.next
            cur = cur.next
        
        if carry:
            cur.next = ListNode(carry)
            carry = 0

        return head.next
